### PR TITLE
made reinstall work on multiple workflows

### PR DESCRIPTION
**Pull request by markgrahamdawson**
closes https://github.com/cylc/cylc-flow/issues/4579

## Updates to code
To make the reinstall operation work for multiple workflows we need to make the main function asynchronous. To do this we **switch out the `reinstall_cli` function which is called in def `main` for a `call_multi` function which is passed `reinstall_cli` wrapped in a `partial`**. 

We then have to go through the call stack and make sure that async and await keywords are used in the correct places - or that async versions of functions are used. So in this case the `reinstall_cli` function needed the `async` keyword added and `get_option_parser` multiworkflow=True.

That would be job done - however `cylc_reinstall` is also imported in `validate_reinstall` so we have to repeat the process of making everything conform with the asynchronous process here also. This includes:

1. using `wrapped_main` instead of `_main`
2. starting a new event loop for` vro_cli`
3. using `parse_id_async` instead of `parse_id`
4. making `vro_cli` asynchronous with the async keyword
5. await `cylc_reinstall` and `scheduler_cli`

As `scheduler_cli` was now being called asynchronously but was not an async function so this needed to be implemented in `cylc/flow/scheduler.py` which was achieved by:

1. switching out `parse_ids` for `parse_ids_async`
2. ...some other stuff

## Updates to tests

-- updates to tests --

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: cylc/flow/scheduler_cli.py [hunk 1]
```diff
@@ -419,9 +424,13 @@ def scheduler_cli(options: 'Values', workflow_id_raw: str) -> None:
     )
 
     # run the workflow
-    ret = asyncio.run(
-        _run(scheduler)
-    )
+    # ret = asyncio.new_event_loop().run_until_complete(_run(scheduler))
```
Comment (oliver-sanders) [inline cylc/flow/scheduler_cli.py:427]:
```suggestion
    # Note: The daemonization messes with asyncio so we have to start a
    # new event loop if detaching
```

File: tests/integration/test_reinstall.py [hunk 2]
```diff
@@ -89,15 +89,17 @@ def one_run(one_src, test_dir, run_dir):
         id=str(w_run_dir.relative_to(run_dir)),
     )
 
+from functools import partial
+from cylc.flow.network.multi import call_multi
```
Comment (oliver-sanders) [inline tests/integration/test_reinstall.py:93]: Can stick these at the top of the file.

**Review by oliver-sanders — COMMENTED** [review]
Looks good.

File: cylc/flow/scheduler_cli.py (hunk 1 above)
Comment (wxtim) [reply cylc/flow/scheduler_cli.py:427]: I agree with Oliver on this.

File: cylc/flow/scripts/validate_reinstall.py [hunk 3]
```diff
@@ -124,16 +125,16 @@ def check_tvars_and_workflow_stopped(
 
 @cli_function(get_option_parser)
 def main(parser: COP, options: 'Values', workflow_id: str):
-    sys.exit(vro_cli(parser, options, workflow_id))
+    sys.exit(asyncio.run(vro_cli(parser, options, workflow_id)))
 
 
-def vro_cli(parser: COP, options: 'Values', workflow_id: str):
+async def vro_cli(parser: COP, options: 'Values', workflow_id: str):
```
Comment (wxtim) [inline cylc/flow/scripts/validate_reinstall.py:131]: Tangential (and my fault), but while we're here can we change this for consistency - VRO is dead as a name.
```suggestion
async def vr_cli(parser: COP, options: 'Values', workflow_id: str):
```

**Comment by wxtim** [comment]
I don't think that VR is quite right, sorry:

```console
> # for name in foo foot footling fool; do cylc vip ~/cylc-src/simplest/ -n $name; done
> cylc stop '*'
fool/run1
Command submitted; the scheduler will log any problems.
...
>  # ensure it's stopped
> cylc scan
> # Make some reinstall-worthy change:
> sed -i 's@a => b & c => d@a => b@' ~/cylc-src/simplest/flow.cylc 
> cylc scan
> cylc vr 'foo*'
WorkflowFilesError: invalid workflow name 'foo*' - can only contain: alphanumeric, `/`, `_`, `+`, `-`, `.`, `@`
```

**Comment by markgrahamdawson** [comment]
> I don't think that VR is quite right, sorry - I don't think it will yet take wild cards:
> 
> ```shell
> # for name in foo foot footling fool; do cylc vip ~/cylc-src/simplest/ -n $name; done
> cylc stop '*'
> fool/run1
> Command submitted; the scheduler will log any problems.
>  ...
> # ensure it's stopped
> cylc scan
> # Make some reinstall-worthy change:
> sed -i 's@a => b & c => d@a => b@' ~/cylc-src/simplest/flow.cylc 
> cylc vr 'foo*'
> WorkflowFilesError: invalid workflow name 'foo*' - can only contain: alphanumeric, `/`, `_`, `+`, `-`, `.`, `@`
> ```

Hmm I think this must be something to do with parsing of workflow ids but Im not sure why

**Comment by oliver-sanders** [comment]
@wxtim, this PR is just about `cylc reinstall`, `cylc vr` does not yet accept multiple workflows.

**Comment by wxtim** [comment]
> @wxtim, this PR is just about `cylc reinstall`, `cylc vr` does not yet accept multiple workflows.

So it's out of scope. Agreed.

**Review by oliver-sanders — APPROVED** [review]

**Comment by oliver-sanders** [comment]
LGTM, works well.

Tested `cylc play` with and without detaching locally and remotely to ensure this logic is still working as expected.

**Comment by oliver-sanders** [comment]
There are some unrelated test failures on this branch which have been fixed on master.

Suggest rebasing this branch to pull in the updates:

```console
$ git fetch upstream
$ git checkout reinstall-multi-workflows
$ git branch reinstall-multi-workflows-safe  # make a safe copy before performing rebase
$ git rebase upstream/master
$ git push -f origin HEAD
```

**Review by wxtim — APPROVED** [review]

File: cylc/flow/scheduler_cli.py [hunk 4]
```diff
@@ -419,9 +424,14 @@ def scheduler_cli(options: 'Values', workflow_id_raw: str) -> None:
...
-        _run(scheduler)
-    )
+    if options.no_detach:
+        ret = await _run(scheduler)
+    else:
+        # Note: The daemonization messes with asyncio so we have to start a
+        # new event loop if detaching
+        ret = asyncio.run(
+            _run(scheduler)
+        )
```
Comment (hjoliver) [inline cylc/flow/scheduler_cli.py:434]: Suspected cause of #6291

**Pull request merged**